File: baml_language/sdks/java/baml_bridge/src/main/java/baml_bridge/BamlFunctionSpec.java
package baml_bridge;

import java.util.concurrent.CompletableFuture;

/**
 * Opaque, engine-owned {@code ai.FunctionSpec<TOut>} capability.
 *
 * <p>The generated {@code Fn_spec(...)} factory obtains this value through the
 * Spec operation on the authored function. All other behavior is expressed as
 * ordinary methods on the capability; no synthetic {@code $parse},
 * {@code $render_prompt}, or {@code $build_request} function is involved.
 */
public final class BamlFunctionSpec<TOut> {
    private static final String[] SELF_NAMES = {"self"};

    private final BamlHandle handle;

    private BamlFunctionSpec(BamlHandle handle) {
        this.handle = handle;
    }

    /** Internal wire-decoder entry point. */
    public static BamlFunctionSpec<?> fromHandle(BamlHandle handle) {
        return new BamlFunctionSpec<>(handle);
    }

    /** Internal wire-encoder entry point. */
    public BamlHandle bamlHandle() {
        return handle;
    }

    @SuppressWarnings("unchecked")
    public TOut call() {
        return (TOut) BamlFfi.callSync("ai.FunctionSpec.call", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public TOut call(Object client, Object on_event) {
        return (TOut)
                BamlFfi.callSync(
                        "ai.FunctionSpec.call",
                        new String[] {"self", "client", "on_event"},
                        new Object[] {this, client, on_event},
                        null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<TOut> call_async() {
        return (CompletableFuture<TOut>) (CompletableFuture<?>)
                BamlFfi.callAsync("ai.FunctionSpec.call", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<TOut> call_async(Object client, Object on_event) {
        return (CompletableFuture<TOut>) (CompletableFuture<?>)
                BamlFfi.callAsync(
                        "ai.FunctionSpec.call",
                        new String[] {"self", "client", "on_event"},
                        new Object[] {this, client, on_event},
                        null);
    }

    @SuppressWarnings("unchecked")
    public TOut parse(String json) {
        return (TOut)
                BamlFfi.callSync(
                        "ai.FunctionSpec.parse",
                        new String[] {"self", "json"},
                        new Object[] {this, json},
                        null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<TOut> parse_async(String json) {
        return (CompletableFuture<TOut>) (CompletableFuture<?>)
                BamlFfi.callAsync(
                        "ai.FunctionSpec.parse",
                        new String[] {"self", "json"},
                        new Object[] {this, json},
                        null);
    }

    public BamlPrompt prompt() {
        return (BamlPrompt)
                BamlFfi.callSync(
                        "ai.FunctionSpec.prompt", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<BamlPrompt> prompt_async() {
        return (CompletableFuture<BamlPrompt>) (CompletableFuture<?>)
                BamlFfi.callAsync(
                        "ai.FunctionSpec.prompt", SELF_NAMES, new Object[] {this}, null);
    }

    public Object build_request() {
        return BamlFfi.callSync("ai.FunctionSpec.build_request", SELF_NAMES, new Object[] {this}, null);
    }

    public Object build_request(Object client) {
        return BamlFfi.callSync(
                "ai.FunctionSpec.build_request",
                new String[] {"self", "client"},
                new Object[] {this, client},
                null);
    }

    public CompletableFuture<Object> build_request_async() {
        return BamlFfi.callAsync(
                "ai.FunctionSpec.build_request", SELF_NAMES, new Object[] {this}, null);
    }

    public CompletableFuture<Object> build_request_async(Object client) {
        return BamlFfi.callAsync(
                "ai.FunctionSpec.build_request",
                new String[] {"self", "client"},
                new Object[] {this, client},
                null);
    }

    public String name() {
        return (String) BamlFfi.callSync("ai.FunctionSpec.name", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<String> name_async() {
        return (CompletableFuture<String>) (CompletableFuture<?>)
                BamlFfi.callAsync("ai.FunctionSpec.name", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public java.util.Map<String, Object> arguments() {
        return (java.util.Map<String, Object>)
                BamlFfi.callSync("ai.FunctionSpec.arguments", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<java.util.Map<String, Object>> arguments_async() {
        return (CompletableFuture<java.util.Map<String, Object>>) (CompletableFuture<?>)
                BamlFfi.callAsync("ai.FunctionSpec.arguments", SELF_NAMES, new Object[] {this}, null);
    }

    public BamlType output_type() {
        return (BamlType)
                BamlFfi.callSync(
                        "ai.FunctionSpec.output_type", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<BamlType> output_type_async() {
        return (CompletableFuture<BamlType>) (CompletableFuture<?>)
                BamlFfi.callAsync(
                        "ai.FunctionSpec.output_type", SELF_NAMES, new Object[] {this}, null);
    }

    public Object tools() {
        return BamlFfi.callSync("ai.FunctionSpec.tools", SELF_NAMES, new Object[] {this}, null);
    }

    public CompletableFuture<Object> tools_async() {
        return BamlFfi.callAsync("ai.FunctionSpec.tools", SELF_NAMES, new Object[] {this}, null);
    }

    public String client_id() {
        return (String)
                BamlFfi.callSync("ai.FunctionSpec.client_id", SELF_NAMES, new Object[] {this}, null);
    }

    @SuppressWarnings("unchecked")
    public CompletableFuture<String> client_id_async() {
        return (CompletableFuture<String>) (CompletableFuture<?>)
                BamlFfi.callAsync("ai.FunctionSpec.client_id", SELF_NAMES, new Object[] {this}, null);
    }
}

File: baml_language/sdks/csharp/bridge_csharp/src/Values/BamlFunctionSpec.cs
using Baml.Generated.V1;
using System.Collections.ObjectModel;

namespace Baml;

/// <summary>
/// An opaque, bound <c>ai.FunctionSpec&lt;TFinal&gt;</c> capability.
/// </summary>
public sealed class BamlFunctionSpec<TFinal> : IDisposable
{
    private readonly BamlGeneratedRegistry registry;
    private readonly TypeDeclaration<TFinal> finalType;
    private BamlHandle? handle;

    internal BamlFunctionSpec(
        BamlHandle handle,
        BamlGeneratedRegistry registry,
        TypeDeclaration<TFinal> finalType)
    {
        this.handle = handle ?? throw new ArgumentNullException(nameof(handle));
        this.registry = registry ?? throw new ArgumentNullException(nameof(registry));
        this.finalType = finalType ?? throw new ArgumentNullException(nameof(finalType));
    }

    /// <summary>Execute this bound spec using its default client.</summary>
    public TFinal Call(CancellationToken cancellationToken = default) =>
        CallAsync(cancellationToken).GetAwaiter().GetResult();

    /// <summary>Execute this bound spec asynchronously using its default client.</summary>
    public async Task<TFinal> CallAsync(
        CancellationToken cancellationToken = default)
    {
        BamlGeneratedValue result = await InvokeAsync(
                "ai.FunctionSpec.call",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false);
        return registry.Decode(finalType, result);
    }

    /// <summary>Parse an existing model reply using this spec's output type.</summary>
    public TFinal Parse(
        string json,
        CancellationToken cancellationToken = default) =>
        ParseAsync(json, cancellationToken).GetAwaiter().GetResult();

    /// <summary>Parse an existing model reply asynchronously.</summary>
    public async Task<TFinal> ParseAsync(
        string json,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(json);
        BamlGeneratedValue result = await InvokeAsync(
                "ai.FunctionSpec.parse",
                [new("json", BamlGeneratedValue.CreateString(json))],
                cancellationToken)
            .ConfigureAwait(false);
        return registry.Decode(finalType, result);
    }

    /// <summary>Render the portable provider-neutral prompt for this spec.</summary>
    public BamlPrompt Prompt(CancellationToken cancellationToken = default) =>
        PromptAsync(cancellationToken).GetAwaiter().GetResult();

    /// <summary>Render the portable prompt asynchronously.</summary>
    public async Task<BamlPrompt> PromptAsync(
        CancellationToken cancellationToken = default)
    {
        BamlGeneratedValue result = await InvokeAsync(
                "ai.FunctionSpec.prompt",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false);
        return new BamlPrompt(result.ReadPromptAst(), registry);
    }

    /// <summary>Build the provider HTTP request without invoking the model.</summary>
    public BamlValue BuildRequest(CancellationToken cancellationToken = default) =>
        BuildRequestAsync(cancellationToken).GetAwaiter().GetResult();

    /// <summary>Build the provider HTTP request asynchronously.</summary>
    public async Task<BamlValue> BuildRequestAsync(
        CancellationToken cancellationToken = default) =>
        new(await InvokeAsync(
                "ai.FunctionSpec.build_request",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false));

    public string Name(CancellationToken cancellationToken = default) =>
        NameAsync(cancellationToken).GetAwaiter().GetResult();

    public async Task<string> NameAsync(
        CancellationToken cancellationToken = default) =>
        (await InvokeAsync(
                "ai.FunctionSpec.name",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false)).ReadString();

    public IReadOnlyDictionary<string, BamlValue> Arguments(
        CancellationToken cancellationToken = default) =>
        ArgumentsAsync(cancellationToken).GetAwaiter().GetResult();

    public async Task<IReadOnlyDictionary<string, BamlValue>> ArgumentsAsync(
        CancellationToken cancellationToken = default)
    {
        BamlGeneratedValue result = await InvokeAsync(
                "ai.FunctionSpec.arguments",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false);
        return new ReadOnlyDictionary<string, BamlValue>(
            result.ReadMapEntries().ToDictionary(
                pair => pair.Key,
                pair => new BamlValue(pair.Value),
                StringComparer.Ordinal));
    }

    public BamlType OutputType(CancellationToken cancellationToken = default) =>
        OutputTypeAsync(cancellationToken).GetAwaiter().GetResult();

    public async Task<BamlType> OutputTypeAsync(
        CancellationToken cancellationToken = default) =>
        (await InvokeAsync(
                "ai.FunctionSpec.output_type",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false)).ReadType();

    /// <summary>
    /// Return the spec's toolbox as a type-erased BAML value. Tool schemas and
    /// callbacks remain represented by their canonical BAML values.
    /// </summary>
    public BamlValue Tools(CancellationToken cancellationToken = default) =>
        ToolsAsync(cancellationToken).GetAwaiter().GetResult();

    public async Task<BamlValue> ToolsAsync(
        CancellationToken cancellationToken = default) =>
        new(await InvokeAsync(
                "ai.FunctionSpec.tools",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false));

    public string ClientId(CancellationToken cancellationToken = default) =>
        ClientIdAsync(cancellationToken).GetAwaiter().GetResult();

    public async Task<string> ClientIdAsync(
        CancellationToken cancellationToken = default) =>
        (await InvokeAsync(
                "ai.FunctionSpec.client_id",
                additionalArguments: null,
                cancellationToken)
            .ConfigureAwait(false)).ReadString();

    public void Dispose()
    {
        BamlHandle? current = Interlocked.Exchange(ref handle, null);
        current?.Dispose();
    }

    public override string ToString() => "<BamlFunctionSpec>";

    internal BamlHandle Handle =>
        Volatile.Read(ref handle)
        ?? throw new ObjectDisposedException(nameof(BamlFunctionSpec<TFinal>));

    internal void RequireRegistry(BamlGeneratedRegistry expected)
    {
        _ = Handle;
        if (!ReferenceEquals(registry, expected))
        {
            throw new InvalidOperationException(
                "The BAML FunctionSpec belongs to another generated program.");
        }
    }

    private Task<BamlGeneratedValue> InvokeAsync(
        string functionIdentity,
        IReadOnlyList<KeyValuePair<string, BamlGeneratedValue>>? additionalArguments,
        CancellationToken cancellationToken)
    {
        var arguments = new List<KeyValuePair<string, BamlGeneratedValue>>(
            1 + (additionalArguments?.Count ?? 0))
        {
            new("self", BamlGeneratedValue.CreateHandle(Handle)),
        };
        if (additionalArguments is not null)
        {
            arguments.AddRange(additionalArguments);
        }

        return registry.RequireProgram().CallRuntimeMethodAsync(
            functionIdentity,
            arguments.AsReadOnly(),
            cancellationToken);
    }
}
